ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,363 @@
+    }
+    long sizeToBeWritten = Math.min(size, capacityInBytes - offset);
+    long sizeLeft = size - sizeToBeWritten;
+    long bytesWritten = 0;
+    while (bytesWritten < sizeToBeWritten) {
+      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, sizeToBeWritten - bytesWritten);
+    }
+    if (offset + sizeToBeWritten > endOffset.get()) {
+      endOffset.set(offset + sizeToBeWritten);
+    }
+    if (sizeLeft > 0) {
+      next.writeFrom(channel, 0, sizeLeft);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * <p/>
+   * If {@code position} + {@code buffer.remaining()} exceeds the capacity of the current segment, bytes are read
+   * starting from {@code position} until capacity is reached. The rest of the bytes upto {@code buffer.remaining()} are
+   * read from the next segment (if one exists). A read is not started if it cannot be completed.
+   * <p/>
+   * Using this function does not cause any state changes
+   * @param buffer The buffer into which the read needs to write to
+   * @param position The position to start the read from
+   * @throws IllegalArgumentException if {@code position} < 0 or if {@code buffer} size is greater than the data
+   * available for read.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   * @throws UnsupportedOperationException if this LogSegment is intended to be swap space or if it is supposed to free
+   */
+  @Override
+  public void readInto(ByteBuffer buffer, long position)
+      throws IOException {
+    if (position < 0 || position > getEndOffset()) {
+      throw new IllegalArgumentException(
+          "Provided position [" + position + "] is out of bounds for the segment [" + file.getAbsolutePath()
+              + "] with end offset [" + getEndOffset() + "]");
+    }
+    if (position + buffer.remaining() > getEndOffset()) {
+      logger.info("{} : Trying to read from position {} into a buffer of size {} when end offset is {}",
+          file.getAbsolutePath(), position, buffer.remaining(), getEndOffset());
+      ensureNext(metrics.overflowReadError);
+    }
+    long sizeToBeRead = Math.min(buffer.remaining(), getEndOffset() - position);
+    long sizeLeft = buffer.remaining() - sizeToBeRead;
+    long bytesRead = 0;
+    int savedLimit = buffer.limit();
+    buffer.limit(buffer.position() + (int) sizeToBeRead);
+    try {
+      while (bytesRead < sizeToBeRead) {
+        bytesRead += fileChannel.read(buffer, position + bytesRead);
+      }
+    } finally {
+      buffer.limit(savedLimit);
+    }
+    if (sizeLeft > 0) {
+      next.readInto(buffer, 0);
+    }
+  }
+
+  /**
+   * Gets the {@link File} and {@link FileChannel} backing this log segment. Also increments a ref count.
+   * <p/>
+   * The expectation is that a matching {@link #closeView()} will be called eventually to decrement the ref count.
+   * @param offset the offset that will used to start the read operation from.
```

Comment:
If we are trying to mimic getView() in current Log, then isn't the api should look like getView(List<BlobReadOptions>). But then looks like you have designed the api with the intention to create the StoreMessageReadSet in Log and not within the logSegment per se. So, I don't see a need for an offset here. Anyways, lets try to sync up offline today regarding this.